### PR TITLE
GraphicsMagick >=1.3.22 compability

### PR DESCRIPTION
**Pull request by norrs**
Fixes https://github.com/sharneng/gm4java/issues/9

### CODE REVIEW

**Comment by kennethxu** [comment]
Thanks for the code contribution especially the fix for Java 8. 

It was my mistake to make safe-mode by default instead of an opt-in configuration. Your change works but I'm afraid we are deepening my mistake further into the API.

I would like the client to be able to use it without worrying about this in the future. One possible solution is to make the call to "gm version" to detect which gm we are working with and use appropriate command. 

If that is too hard, I would rather just get rid of safe-mode all together and post a warning that if someone intend to use new version of gm4java with old version of gm, make sure he/she does intensive test.

Honestly, the old version may just work fine without safe mode. Having safe mode was purely due to my laziness. When I release gm4java, I didn't have time to full test the gm out of safe mode.

Thoughts?

**Comment by norrs** [comment]
@kennethxu: I guess we could call gm version to indicate version yes. I'll look at rebasing the last commit tonight and see if I can do some changes to make it fetch the version from "gm version" and clean up. 

I'd rather avoid breaking backwards compatibility, as long as we can detect version there should be no need to nag the user about having a new gm4java when using a old version of gm as long as we provide the same flags that's always been in use. I also this is wise so we don't break applications that depends on gm4java. (and still use an old gm) 

AFAIK safe-mode does something to init and destroy for releasing resources. Assumptions is that it is done to ensure that a state from one "image" doesn't get transferred to the other "Image" in batch processing.

**Comment by norrs** [comment]
@kennethxu : Updated PR, hope this works.

NB: it does break backwards compatibility if it fails to execute the "gm version" with throwing an runtime exception, hope that is okey. (as it shouldn't happen!) 

Would also be nice if you could fix #12 to ensure this works under the platform Windows.

**Comment by norrs** [comment]
Wait with merging this one please, tests were running fine in IDE but no in command line. 
I'll investigate…

File: src/test/java/org/gm4java/engine/support/GMConnectionPoolTest.java [hunk 1]
```diff
@@ -78,7 +81,8 @@ public void setup() throws Exception {
             processes[i] = mock(ReaderWriterProcess.class);
         }
         when(factory.getProcess(Matchers.<String[]> anyVararg())).thenReturn(process);
-        sut = new GMConnectionPool(config);
+        when(commandSelector.gmCommand()).thenReturn(Constants.GM_COMMAND.clone());
```
Comment (norrs) [inline src/test/java/org/gm4java/engine/support/GMConnectionPoolTest.java]: This one was important to ensure that I clone() :-)

**Comment by norrs** [comment]
@kennethxu : Updated, you can now review it.

(Sorry for the spam.)

**Comment by norrs** [comment]
If you need anything else @kennethxu , leave a comment and tag my username. 

We don't have this in production, but we're using a local build which is used in a local dev/stage environment until rest of our features are ready to be launched. 

I'll leave comments here if we find any issues if the PR is still open or create new issues.

@norrs out.

**Comment by PavelCibulka** [comment]
@kennethxu It would be really helpful, if you can integrate this and release maven artifact.

**Comment by mhurne** [comment]
Any chance of this being merged and released in the near future? We have considered using an older version of graphicsmagick to work around this incompatibility, but now that it has been revealed that there are some significant security vulnerabilities to be addressed, we aren't comfortable with that workaround.

**Comment by mhurne** [comment]
Thanks, @kennethxu !

**Comment by kennethxu** [comment]
@mhurne and others, sorry for taking this long, I'm no longer working on a project that request this lib and I have been really busy recently.

After merging, I noticed that the API was changed as a result. Over the weekend, I force myself to get sometime to look into this. I'm making some change to make sure the public API stays the same. I have also refactor the code so that the version detector mechanism gives gm4java the ability know the detail of the GM installed. Now used for the purpose of command line but is flexible for future expansion.

**Comment by kennethxu** [comment]
Change was committed, would you mind to pull the and test?

**Comment by mhurne** [comment]
I've asked the appropriate engineer on my team to test the changes for our use case, but don't believe it has happened yet. I'll let you know how it goes when I hear from him.

**Comment by kennethxu** [comment]
@mhurne, thank you. BTW, I forgot to mention that the SimplGMService was unusable if GraphicsMagick is not on the PATH, because CommandSelector in the constructor alway use the DEFAULT_GM_PATH. Another reason for my refactoring. :)

**Comment by mhurne** [comment]
We've done some testing of these changes with graphicsmagick 1.3.23 and things are looking good!

**Comment by mhurne** [comment]
@kennethxu, do you anticipate releasing these changes a publishing a new build to Maven Central soon?

**Comment by kennethxu** [comment]
@mhurne planning but having some difficulty with my oss.sonatype.org account. Hope to resolve before I go on vacation on 6/2, when I have limited to no access.

**Comment by matthurne** [comment]
Understood. FWIW, Bintray jcenter might be an attractive alternative.

**Comment by kennethxu** [comment]
Released to maven http://mvnrepository.com/artifact/com.sharneng/gm4java/1.1.1

**Comment by matthurne** [comment]
Thanks!

**Pull request merged**